internal/k8s/statefulset/statefulset.go

```diff
@@ -0,0 +1,162 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package statefulset
+
+import (
+	"context"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/k8s"
+	"github.com/vdaas/vald/internal/log"
+	appsv1 "k8s.io/api/apps/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// StatefulSetWatcher is a type alias for k8s resource controller
+type StatefulSetWatcher k8s.ResourceController
+
+type reconciler struct {
+	ctx         context.Context
+	mgr         manager.Manager
+	name        string
+	namespace   string
+	onError     func(err error)
+	onReconcile func(rs map[string][]StatefulSet)
+	pool        sync.Pool
+}
+
+// Statefulset is a type alias for the k8s statefulset definition.
+type StatefulSet = appsv1.StatefulSet
+
+// New returns the StatefulSetWather that implements reconciliation loop, or any error occured.
+func New(opts ...Option) (StatefulSetWatcher, error) {
+	r := &reconciler{
+		pool: sync.Pool{
+			New: func() interface{} {
+				return make(map[string][]StatefulSet)
+			},
+		},
+	}
+
+	for _, opt := range append(defaultOpts, opts...) {
+		if err := opt(r); err != nil {
+			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
+		}
+	}
+
+	return r, nil
+}
+
+// Reconcile implements k8s reconciliation loop to retrive the StatefulSet information from k8s.
+func (r *reconciler) Reconcile(req reconcile.Request) (res reconcile.Result, err error) {
+	ssl := new(appsv1.StatefulSetList)
+
+	err = r.mgr.GetClient().List(r.ctx, ssl)
+	if err != nil {
+		if r.onError != nil {
+			r.onError(err)
+		}
+		res = reconcile.Result{
+			Requeue:      true,
+			RequeueAfter: time.Millisecond * 100,
+		}
+		if k8serrors.IsNotFound(err) {
+			log.Warn(errors.ErrK8sResourceNotFound(err))
+			res.RequeueAfter = time.Second
+			return res, nil
+		}
+		return
+	}
+
+	ssm := r.pool.Get().(map[string][]StatefulSet)
+
+	for _, statefulset := range ssl.Items {
+		name, ok := statefulset.GetObjectMeta().GetLabels()["app"]
+		if !ok {
+			pns := strings.Split(statefulset.GetName(), "-")
+			name = strings.Join(pns[:len(pns)-1], "-")
+		}
+		if _, ok := ssm[name]; !ok {
+			ssm[name] = make([]StatefulSet, 0)
+		}
+
+		ssm[name] = append(ssm[name], statefulset)
+	}
+
+	for name := range ssm {
+		l := len(ssm[name])
+		ssm[name] = ssm[name][:l:l]
+	}
+
+	if r.onReconcile != nil {
+		r.onReconcile(ssm)
+	}
+
+	for name := range ssm {
+		ssm[name] = ssm[name][:0]
+	}
```

Comment:
```suggestion
	appList := make(map[string]bool)
	for _, statefulset := range ssl.Items {
		name, ok := statefulset.GetObjectMeta().GetLabels()["app"]
		if !ok {
			pns := strings.Split(statefulset.GetName(), "-")
			name = strings.Join(pns[:len(pns)-1], "-")
		}
		if _, ok := ssm[name]; !ok {
			ssm[name] = make([]StatefulSet, 1)
		}
		if !appList[name] {
			appList[name]=true
		}
		l := len(ssm[name])+1
		ssm[name] = append(ssm[name], statefulset)[:l:l]
	}

	if r.onReconcile != nil {
		r.onReconcile(ssm)
	}

	for name := range ssm {
		if !appList[name] {
			delete(ssm, name)
		}else{
			ssm[name] = ssm[name][:0]
		}

	}
```

Reply:
I am sorry, If I will make a misstake.
I think this code occurs re-alocate when next roop occurs. because the capacity is adjusted with the following code.
> ssm[name] = append(ssm[name], statefulset)[:l:l]

for Is the following correct??

```

for _, statefulset := range ssl.Items {
		.........
		ssm[name] = append(ssm[name], statefulset)
	}
	if r.onReconcile != nil {
		r.onReconcile(ssm)
	}
	for name := range ssm {
		if !appList[name] {
			delete(ssm, name)
		}else{
                        l = len(ssm[name])
			ssm[name] = ssm[name][:l:l]
			ssm[name] = ssm[name][:0]
		}
	}

```

Reply:
your code is not correct, but I understand we should fix
`ssm[name] = append(ssm[name], statefulset)[:l:l]`

Reply:
updated

Reply:
but, I don't know the meaning of this line
```
ssm[name][:l:l]
```
do we need fix capacity???

Reply:
for pool we already seek pointer to zero, which means capacity is always fit to maximum size of k8s resources.I think we don't need that loop.

Reply:
when the number of items is larger than capacity of lists, capacity will be raised enough size (not minimum size).
So, we fix capacity at end of the loop.

Reply:
I know about it, my question is, is there any problem about it???

let you see eg

1st reconcile
A resource has 4 pods

2nd reconcile
A resource has 2 pods

your code
after 2nd reconcile pool capacity becomes 2

my code
after 2nd reconcile pool capacity is still 4 but length is 2

is there any problem about it??

Reply:
yap, there is no clear reason.
we just thought it would be good to secure the necessary and sufficient capacity.

Reply:
hmm

Reply:
@hlts2 what do you think?
it's your question.

Reply:
my suggestion is write this code `ssm[name] = ssm[name][:0:len(ssm[name])]` @ L152

Reply:
and remove capacity loop